asllib/aslspec/latex.ml

```diff
@@ -0,0 +1,267 @@
+let regexp_spaces_segment = Str.regexp "[ \t]+"
+let union_macro_name = "cup"
+
+(** A LaTeX underscore escape sequence. *)
+let escaped_underscore = {|\_|}
+
+let shrink_whitespace str =
+  let regexp_newline_segment = Str.regexp "[ \n\r]+" in
+  Str.global_replace regexp_newline_segment " " str
+
+let escape_underscores str =
+  Str.global_replace regex_underscore escaped_underscore str
+
+let underscore_to_space str = Str.global_replace regex_underscore " " str
+
+type font_type = Text | TextTT | TextSF | TextSC | TextIT
+
+(** [pp_one_arg_macro fmt macro_name pp_arg arg] renders a LaTeX macro named
+    [macro_name] with one argument, applied to the argument is [arg], which is
+    formatted using [pp_arg]. *)
+let pp_one_arg_macro fmt macro_name pp_arg arg =
+  fprintf fmt "\\%s{%a}" macro_name pp_arg arg
+
+(** [pp_macro fmt macro_name] renders a LaTeX macro by prefixing it with a
+    backslash. *)
+let pp_macro fmt macro_name = fprintf fmt "\\%s" macro_name
+
+(** [spec_var_to_latex_var ~font_type var_str] returns a version of the
+    specification variable [var_str] suitable for LaTeX math mode using the
+    specified [font_type]. *)
+let spec_var_to_latex_var ~font_type var_str =
+  if String.equal var_str "_" then "\\Ignore"
+  else
+    let font_macro_name =
+      match font_type with
+      | Text -> "text"
+      | TextTT -> "texttt"
+      | TextSF -> "textsf"
+      | TextSC -> "textsc"
+      | TextIT -> "textit"
+    in
+    Format.asprintf {|%a{%s}|} pp_macro font_macro_name
+      (escape_underscores var_str)
+
+let spec_var_to_prose var_str = underscore_to_space var_str
+let to_math_mode str = "$" ^ str ^ "$"
+let spec_var_to_template_var var_str = "{" ^ var_str ^ "}"
+let remove_underscores str = Str.global_replace (Str.regexp_string "_") "" str
+let elem_name_to_math_macro elem_name = "\\" ^ remove_underscores elem_name
+let pp_comma = fun fmt () -> fprintf fmt ", "
+
+(** [substitute_spec_vars_by_latex_vars math_mode s vars] returns a string [s]
+    with every variable like [my_var] is substituted into [\texttt{my\_var}],
+    which makes the returned string suitable to typsetting with LaTeX. If
+    [math_mode] is true then the result is surrounded by [$$]. *)
+let substitute_spec_vars_by_latex_vars ~math_mode str vars =
+  let module StringMap = Map.Make (String) in
+  let substitutions_map =
+    List.fold_left
+      (fun acc_map var_str ->
+        let latex_var = spec_var_to_latex_var ~font_type:TextTT var_str in
+        let latex_var =
+          if math_mode then to_math_mode latex_var else latex_var
+        in
+        let template_var = spec_var_to_template_var var_str in
+        StringMap.add template_var latex_var acc_map)
+      StringMap.empty vars
+  in
+  let template_var_regexp = Str.regexp "{[a-zA-Z0-9_']+}" in
+  let blocks = Str.full_split template_var_regexp str in
+  List.fold_left
+    (fun acc block ->
+      match block with
+      | Str.Text txt -> acc ^ txt
```

Comment:
consider putting it into a list and using `String.concat` at the end, this way you allocate the string only once.